### ISSUE TITLE
SHACL - Qualified shapes

### ISSUE THREAD

**Issue by hmottestad** [eclipse/rdf4j#2287]
Support for qualified shapes allows users to write things like: A person should only have one MainAddress, but could have other addresses like a SecondaryAddress.

```
ex:QualifiedPersonShape
    a sh:NodeShape ;
    sh:targetClass foaf:Person ;
    sh:property [
        sh:path ex:address ;
        sh:qualifiedValueShape [
            sh:class ex:MainAddress;
        ] ;
        sh:qualifiedMaxCount 1 ;
    ] 
    .
```

**Comment by gauri-b** [comment]
Hi,
I'm getting the error "plugin execution not covered by lifecycle configuration parent " in pom.xml of the shacl folder. How can this be fixed?

**Comment by shubhi0108** [comment]
Hi,

We (@navdhaagarwal and I) would like to work on this issue. We know that the Lucene Sail is used for search queries, and we need to combine the Values clause with the Filter Not Exists. Can you please be a little more specific as to which file we are supposed to look into.

**Comment by hmottestad** [comment]
@gauri-b Can you please create a new issue and describe what command you are running and paste the output you get? This issue is for tracking support for qualified shapes support in the ShaclSail.

**Comment by rdstn** [comment]
Found a use case for this, related to languages. If I want to say that a property has values for all languages in a set, the obvious way to do it is something like this:

```
ex:LanguageShape
	a sh:NodeShape ;
	sh:targetClass ex:Language ;
	sh:property [
		sh:path ex:prop ;
		sh:languageIn ( "en" "fr" ) ;
	] ;
	sh:property [
		sh:path ex:prop ;
		sh:qualifiedValueShape [ sh:languageIn ( "en" ) ] ;
		sh:qualifiedMinCount 1 ;
	] ;
	sh:property [
		sh:path ex:prop ;
		sh:qualifiedValueShape [ sh:languageIn ( "fr" ) ] ;
		sh:qualifiedMinCount 1 ;
	] .
```

This works against the following test data:

```
ex:Obj
    a ex:Language ;
    ex:prop "One"@mi ;
    ex:prop "Two"@en ;
    ex:prop "Three"@fr .
```

I suppose that qualified shapes will need the new AST (#2083) to be implemented?

**Comment by hmottestad** [comment]
I'm not sure if qualified value shapes would need the new AST. I would have to try.

This particular case might also be solvable with a combination of
 - uniqueLang 
 - minCount and maxCount
 - languageIn

**Comment by rdstn** [comment]
The use case can't be resolved by using `uniqueLang` except for instances where `minCount` == `maxCount` == `languageIn` count. 

A counter-example of that would be the name for something. A book must have a name for all languages it is available in. However, it may have multiple names per language, corresponding to different editions or translations.

Something with `sh:and` and `sh:or` could have been an option, if `sh:and` and `sh:or` could be used for constraint components rather than shapes. However, at the moment, the only resolution I can see is with qualified shapes.

**Comment by hmottestad** [comment]
See: https://github.com/eclipse/rdf4j/issues/2858

**Comment by hmottestad** [comment]
This was implemented in RDF4J 3.7.0.
